### ISSUE TITLE
kpt function sdk docker build fails

### ISSUE THREAD

**Issue by artbegolli** [GoogleContainerTools/kpt#684]
This is related to https://github.com/GoogleContainerTools/kpt/issues/672 and outlined in this comment https://github.com/GoogleContainerTools/kpt/issues/672#issuecomment-635668540

Notably the build fails running npm ci step with no discernible error message.

From tweaking with the generated `package.json` and comparing with the examples provided that seemed to build fine - it looks like the error is being caused by the `typescript` version with builds failing on `^3.5.3` but successful on `~3.7.0`. 

Seems as though this should be updated in the sdk.

Stack Trace:
```
Sending build context to Docker daemon  959.5kB
Step 1/14 : FROM node:10-alpine as builder
 ---> a07f309c9753
Step 2/14 : RUN mkdir -p /home/node/app &&     chown -R node:node /home/node/app
 ---> Using cache
 ---> 6f5839478e84
Step 3/14 : USER node
 ---> Using cache
 ---> f571ae2d4482
Step 4/14 : WORKDIR /home/node/app
 ---> Using cache
 ---> 4c84293956e6
Step 5/14 : COPY --chown=node:node package*.json ./
 ---> Using cache
 ---> f744c24b1f7e
Step 6/14 : RUN npm ci
 ---> Running in 7b99026b13e9

> create-kpt-functions@0.16.4 postinstall /home/node/app/node_modules/create-kpt-functions
> go-npm install

Downloading from URL: https://storage.googleapis.com/kpt-functions/v0.16.4/typegen_linux_amd64.tar.gz

> edna-generator@0.1.0 prepare /home/node/app
> npm run build


> edna-generator@0.1.0 build /home/node/app
> tsc

Version 3.8.3
Syntax:   tsc [options] [file...]

Examples: tsc hello.ts
          tsc --outFile file.js file.ts
          tsc @args.txt
          tsc --build tsconfig.json

Options:
 -h, --help                                         Print this message.
 -w, --watch                                        Watch input files.
 --pretty                                           Stylize errors and messages using color and context (experimental).
 --all                                              Show all compiler options.
 -v, --version                                      Print the compiler's version.
 --init                                             Initializes a TypeScript project and creates a tsconfig.json file.
 -p FILE OR DIRECTORY, --project FILE OR DIRECTORY  Compile the project given the path to its configuration file, or to a folder with a 'tsconfig.json'.
 -b, --build                                        Build one or more projects and their dependencies, if out of date
 -t VERSION, --target VERSION                       Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'.
 -m KIND, --module KIND                             Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'.
 --lib                                              Specify library files to be included in the compilation.
                                                      'es5' 'es6' 'es2015' 'es7' 'es2016' 'es2017' 'es2018' 'es2019' 'es2020' 'esnext' 'dom' 'dom.iterable' 'webworker' 'webworker.importscripts' 'scripthost' 'es2015.core' 'es2015.collection' 'es2015.generator' 'es2015.iterable' 'es2015.promise' 'es2015.proxy' 'es2015.reflect' 'es2015.symbol' 'es2015.symbol.wellknown' 'es2016.array.include' 'es2017.object' 'es2017.sharedmemory' 'es2017.string' 'es2017.intl' 'es2017.typedarrays' 'es2018.asyncgenerator' 'es2018.asynciterable' 'es2018.intl' 'es2018.promise' 'es2018.regexp' 'es2019.array' 'es2019.object' 'es2019.string' 'es2019.symbol' 'es2020.bigint' 'es2020.promise' 'es2020.string' 'es2020.symbol.wellknown' 'esnext.array' 'esnext.symbol' 'esnext.asynciterable' 'esnext.intl' 'esnext.bigint'
 --allowJs                                          Allow javascript files to be compiled.
 --jsx KIND                                         Specify JSX code generation: 'preserve', 'react-native', or 'react'.
 -d, --declaration                                  Generates corresponding '.d.ts' file.
 --declarationMap                                   Generates a sourcemap for each corresponding '.d.ts' file.
 --sourceMap                                        Generates corresponding '.map' file.
 --outFile FILE                                     Concatenate and emit output to single file.
 --outDir DIRECTORY                                 Redirect output structure to the directory.
 --removeComments                                   Do not emit comments to output.
 --noEmit                                           Do not emit outputs.
 --strict                                           Enable all strict type-checking options.
 --noImplicitAny                                    Raise error on expressions and declarations with an implied 'any' type.
 --strictNullChecks                                 Enable strict null checks.
 --strictFunctionTypes                              Enable strict checking of function types.
 --strictBindCallApply                              Enable strict 'bind', 'call', and 'apply' methods on functions.
 --strictPropertyInitialization                     Enable strict checking of property initialization in classes.
 --noImplicitThis                                   Raise error on 'this' expressions with an implied 'any' type.
 --alwaysStrict                                     Parse in strict mode and emit "use strict" for each source file.
 --noUnusedLocals                                   Report errors on unused locals.
 --noUnusedParameters                               Report errors on unused parameters.
 --noImplicitReturns                                Report error when not all code paths in function return a value.
 --noFallthroughCasesInSwitch                       Report errors for fallthrough cases in switch statement.
 --types                                            Type declaration files to be included in compilation.
 --esModuleInterop                                  Enables emit interoperability between CommonJS and ES Modules via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'.
 @<file>                                            Insert command line options and files from a file.
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! edna-generator@0.1.0 build: `tsc`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the edna-generator@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_cacache/_logs/2020-05-28T23_23_44_412Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! edna-generator@0.1.0 prepare: `npm run build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the edna-generator@0.1.0 prepare script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/node/.npm/_logs/2020-05-28T23_23_44_430Z-debug.log
The command '/bin/sh -c npm ci' returned a non-zero code: 1
Error: Failed to build docker image
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cmd/docker_build.js:32:19
    at Object.processDockerfile (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cmd/docker_create.js:66:9)
    at Object.dockerBuild (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cmd/docker_build.js:23:21)
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:112:32
    at Generator.next (<anonymous>)
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:24:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:20:12)
    at main (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:37:12)
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:124:15
    at Generator.next (<anonymous>)
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:24:71
    at new Promise (<anonymous>)
    at __awaiter (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:20:12)
    at /Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:122:8
    at Object.<anonymous> (/Users/morgantep/code/google/pso/mp/mp-data-automation/generator/node_modules/create-kpt-functions/dist/cli.js:130:4)
    at Module._compile (internal/modules/cjs/loader.js:1144:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1164:10)
    at Module.load (internal/modules/cjs/loader.js:993:32)
    at Function.Module._load (internal/modules/cjs/loader.js:892:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
    at internal/main/run_main_module.js:17:47
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! edna-generator@0.1.0 kpt:docker-build: `kpt docker-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the edna-generator@0.1.0 kpt:docker-build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/morgantep/.npm/_logs/2020-05-28T23_23_44_709Z-debug.log
```

**Comment by prachirp** [comment]
TL;DR: I repro'd this error locally on [init-package](https://github.com/GoogleContainerTools/kpt-functions-sdk/tree/master/ts/init-package). The problem here is definitely with in the Dockerfile but I couldn't fix it.

I successfully ran `node dist/my_func_run.js --help`. I also repro'd the issue by running `docker build -t gcr.io/kpt-functions-demo/my-func:dev -f build/my_func.Dockerfile .` which means this is an issue with the Dockerfile, not with create-kpt-functions docker-build command. I then updated the base image of the Dockerfile to use node:12-alpine or node:lts-alpine as well as updated outdated versions in package.json but nothing worked.

Debug log:
```
cat /Users/prachirp/.npm/_logs/2020-06-03T04_06_26_215Z-debug.log
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/local/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'run',
1 verbose cli   'kpt:docker-build'
1 verbose cli ]
2 info using npm@6.14.4
3 info using node@v12.18.0
4 verbose run-script [ 'prekpt:docker-build', 'kpt:docker-build', 'postkpt:docker-build' ]
5 info lifecycle init-package@0.1.0~prekpt:docker-build: init-package@0.1.0
6 info lifecycle init-package@0.1.0~kpt:docker-build: init-package@0.1.0
7 verbose lifecycle init-package@0.1.0~kpt:docker-build: unsafe-perm in lifecycle true
8 verbose lifecycle init-package@0.1.0~kpt:docker-build: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/prachirp/workspace/kpt-functions-sdk/ts/init-package/node_modules/.bin:/Users/prachirp/google-cloud-sdk/bin:/Users/prachirp/.istioctl/bin:/Users/prachirp/bin:/usr/local/bin:/usr/local/git/current/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/go/bin
9 verbose lifecycle init-package@0.1.0~kpt:docker-build: CWD: /Users/prachirp/workspace/kpt-functions-sdk/ts/init-package
10 silly lifecycle init-package@0.1.0~kpt:docker-build: Args: [ '-c', 'kpt docker-build' ]
11 silly lifecycle init-package@0.1.0~kpt:docker-build: Returned: code: 1  signal: null
12 info lifecycle init-package@0.1.0~kpt:docker-build: Failed to exec kpt:docker-build script
13 verbose stack Error: init-package@0.1.0 kpt:docker-build: `kpt docker-build`
13 verbose stack Exit status 1
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:315:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:315:20)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
14 verbose pkgid init-package@0.1.0
15 verbose cwd /Users/prachirp/workspace/kpt-functions-sdk/ts/init-package
16 verbose Darwin 19.5.0
17 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "kpt:docker-build"
18 verbose node v12.18.0
19 verbose npm  v6.14.4
20 error code ELIFECYCLE
21 error errno 1
22 error init-package@0.1.0 kpt:docker-build: `kpt docker-build`
22 error Exit status 1
23 error Failed at the init-package@0.1.0 kpt:docker-build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

@frankfarzan any ideas?

**Comment by artbegolli** [comment]
> TL;DR: I repro'd this error locally on [init-package](https://github.com/GoogleContainerTools/kpt-functions-sdk/tree/master/ts/init-package). The problem here is definitely with in the Dockerfile but I couldn't fix it.
> 
> I successfully ran `node dist/my_func_run.js --help`. I also repro'd the issue by running `docker build -t gcr.io/kpt-functions-demo/my-func:dev -f build/my_func.Dockerfile .` which means this is an issue with the Dockerfile, not with create-kpt-functions docker-build command. I then updated the base image of the Dockerfile to use node:12-alpine or node:lts-alpine as well as updated outdated versions in package.json but nothing worked.
> 
> Debug log:
> 
> ```
> cat /Users/prachirp/.npm/_logs/2020-06-03T04_06_26_215Z-debug.log
> 0 info it worked if it ends with ok
> 1 verbose cli [
> 1 verbose cli   '/usr/local/bin/node',
> 1 verbose cli   '/usr/local/bin/npm',
> 1 verbose cli   'run',
> 1 verbose cli   'kpt:docker-build'
> 1 verbose cli ]
> 2 info using npm@6.14.4
> 3 info using node@v12.18.0
> 4 verbose run-script [ 'prekpt:docker-build', 'kpt:docker-build', 'postkpt:docker-build' ]
> 5 info lifecycle init-package@0.1.0~prekpt:docker-build: init-package@0.1.0
> 6 info lifecycle init-package@0.1.0~kpt:docker-build: init-package@0.1.0
> 7 verbose lifecycle init-package@0.1.0~kpt:docker-build: unsafe-perm in lifecycle true
> 8 verbose lifecycle init-package@0.1.0~kpt:docker-build: PATH: /usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/Users/prachirp/workspace/kpt-functions-sdk/ts/init-package/node_modules/.bin:/Users/prachirp/google-cloud-sdk/bin:/Users/prachirp/.istioctl/bin:/Users/prachirp/bin:/usr/local/bin:/usr/local/git/current/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/go/bin
> 9 verbose lifecycle init-package@0.1.0~kpt:docker-build: CWD: /Users/prachirp/workspace/kpt-functions-sdk/ts/init-package
> 10 silly lifecycle init-package@0.1.0~kpt:docker-build: Args: [ '-c', 'kpt docker-build' ]
> 11 silly lifecycle init-package@0.1.0~kpt:docker-build: Returned: code: 1  signal: null
> 12 info lifecycle init-package@0.1.0~kpt:docker-build: Failed to exec kpt:docker-build script
> 13 verbose stack Error: init-package@0.1.0 kpt:docker-build: `kpt docker-build`
> 13 verbose stack Exit status 1
> 13 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
> 13 verbose stack     at EventEmitter.emit (events.js:315:20)
> 13 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
> 13 verbose stack     at ChildProcess.emit (events.js:315:20)
> 13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
> 13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
> 14 verbose pkgid init-package@0.1.0
> 15 verbose cwd /Users/prachirp/workspace/kpt-functions-sdk/ts/init-package
> 16 verbose Darwin 19.5.0
> 17 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "run" "kpt:docker-build"
> 18 verbose node v12.18.0
> 19 verbose npm  v6.14.4
> 20 error code ELIFECYCLE
> 21 error errno 1
> 22 error init-package@0.1.0 kpt:docker-build: `kpt docker-build`
> 22 error Exit status 1
> 23 error Failed at the init-package@0.1.0 kpt:docker-build script.
> 23 error This is probably not a problem with npm. There is likely additional logging output above.
> 24 verbose exit [ 1, true ]
> ```
> 
> @frankfarzan any ideas?

And you had no luck updating both the typescript version to 3.7.0 and kpt-functions to 0.14.0? This seemed to work for me. I'm inclined to think the issue is not with the Dockerfile but with the dependencies instead seeing as the example functions are building fine with a pretty much identical Dockerfile.

**Comment by frankfarzan** [comment]
Fix:

```
--- a/build/generate_tf.Dockerfile
+++ b/build/generate_tf.Dockerfile
@@ -9,7 +9,7 @@ WORKDIR /home/node/app

 # Install dependencies and cache them.
 COPY --chown=node:node package*.json ./
-RUN npm ci
+RUN npm ci --ignore-scripts
```

`npm ci` calls `prepare` hook which calls `tsc`. At this point in Dockerfile, there's no `tsconfig` file. I think previously, `tsc` would just do a NO OP if there was no `tsconfig`, but in a recent (or oder) versions this behavior is different and it fails. `--ignore-scripts` avoids doing a build.

@prachirp can you follow up with making this change to the CLI and making a release?

**Comment by morgante** [comment]
Thanks @frankfarzan, that fixed it.

**Comment by prachirp** [comment]
Fixed with latest release of create-kpt-functions@0.17.0
